File: static/cpp/java/src/main/java/com/epi/OnlineMedian.java
package com.epi;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Random;
import java.util.Scanner;

public class OnlineMedian {
  private static List<Double> globalResult = new ArrayList<>();

  // @include
  private static final int DEFAULT_INITIAL_CAPACITY = 16;

  public static void onlineMedian(Iterator<Integer> sequence) {
    // minHeap stores the larger half seen so far.
    PriorityQueue<Integer> minHeap = new PriorityQueue<>();
    // maxHeap stores the smaller half seen so far.
    PriorityQueue<Integer> maxHeap = new PriorityQueue<>(
        DEFAULT_INITIAL_CAPACITY, Collections.reverseOrder());

    while (sequence.hasNext()) {
      int x = sequence.next();
      if (minHeap.isEmpty()) {
        // This is the very first element.
        minHeap.add(x);
      } else {
        if (x >= minHeap.peek()) {
          minHeap.add(x);
        } else {
          maxHeap.add(x);
        }
      }
      // Ensure minHeap and maxHeap have equal number of elements if
      // an even number of elements is read; otherwise, minHeap must have
      // one more element than maxHeap.
      if (minHeap.size() > maxHeap.size() + 1) {
        maxHeap.add(minHeap.remove());
      } else if (maxHeap.size() > minHeap.size()) {
        minHeap.add(maxHeap.remove());
      }

      // @exclude
      globalResult.add((minHeap.size() == maxHeap.size()
                            ? 0.5 * (minHeap.peek() + maxHeap.peek())
                            : minHeap.peek()));
      // @include
      System.out.println(minHeap.size() == maxHeap.size()
                             ? 0.5 * (minHeap.peek() + maxHeap.peek())
                             : minHeap.peek());
    }
  }
  // @exclude

  private static void smallTest() {
    List<Integer> stream = Arrays.asList(5, 4, 3, 2, 1);
    onlineMedian(stream.iterator());
    List<Double> golden = Arrays.asList(5.0, 4.5, 4.0, 3.5, 3.0);
    assert golden.equals(globalResult);

    globalResult.clear();
    stream = Arrays.asList(1, 2, 3, 4, 5);
    onlineMedian(stream.iterator());
    golden = Arrays.asList(1.0, 1.5, 2.0, 2.5, 3.0);
    assert golden.equals(globalResult);

    globalResult.clear();
    stream = Arrays.asList(1, 0, 3, 5, 2, 0, 1);
    onlineMedian(stream.iterator());
    golden = Arrays.asList(1.0, 0.5, 1.0, 2.0, 2.0, 1.5, 1.0);
    assert golden.equals(globalResult);

    globalResult.clear();
    stream = Arrays.asList(-1);
    onlineMedian(stream.iterator());
    golden = Arrays.asList(-1.0);
    assert golden.equals(globalResult);
  }

  public static void main(String[] args) {
    smallTest();
    Random r = new Random();
    int num;
    if (args.length == 1) {
      num = Integer.parseInt(args[0]);
    } else {
      num = r.nextInt(100000) + 1;
    }
    List<Integer> stream = new ArrayList<>();
    for (int i = 0; i < num; ++i) {
      stream.add(r.nextInt(10000) + 1);
    }
    onlineMedian(stream.iterator());
  }
}

File: static/python/online_median.py
import sys
import random
import heapq

global_result = []


# @include
def online_median(sequence):
    # min_heap stores the larger half seen so far.
    min_heap = []
    # max_heap stores the smaller half seen so far.
    # values in max_heap are negative
    max_heap = []

    for x in sequence:
        heapq.heappush(max_heap, -heapq.heappushpop(min_heap, x))
        # Ensure min_heap and max_heap have equal number of elements if an even
        # number of elements is read; otherwise, min_heap must have one more
        # element than max_heap.
        if len(max_heap) > len(min_heap):
            heapq.heappush(min_heap, -heapq.heappop(max_heap))

        # @exclude
        global_result.append(0.5 * (min_heap[0] + (-max_heap[0]))
                             if len(min_heap) == len(max_heap) else min_heap[0])
        # @include
        print(0.5 * (min_heap[0] + (-max_heap[0]))
              if len(min_heap) == len(max_heap) else min_heap[0])
# @exclude


def small_test():
    stream = [5, 4, 3, 2, 1]
    online_median(iter(stream))
    assert global_result == [5, 4.5, 4, 3.5, 3]

    global_result.clear()
    stream = [1, 2, 3, 4, 5]
    online_median(iter(stream))
    assert global_result == [1, 1.5, 2, 2.5, 3]

    global_result.clear()
    stream = [1, 0, 3, 5, 2, 0, 1]
    online_median(iter(stream))
    assert global_result == [1, 0.5, 1, 2, 2, 1.5, 1]

    global_result.clear()
    stream = [-1]
    online_median(iter(stream))
    assert global_result == [-1]


def main():
    small_test()
    num = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 100000)
    stream = [random.randint(1, 10000) for _ in range(num)]
    print(*stream, sep='\n')
    print()
    online_median(iter(stream))


if __name__ == '__main__':
    main()
